wsrep_tls_service.h

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright (C) 2020 Codership Oy <info@codership.com>
+ *
+ * This file is part of wsrep-API.
+ *
+ * Wsrep-API is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Wsrep-API is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with wsrep-API.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/** @file wsrep_tls_service.h
+ *
+ * This file defines interface for TLS services provided by the application,
+ * used by the provider.
+ *
+ * In order to support both synchronous and asynchronous IO operations,
+ * the interface is designed to work with sockets in both blocking
+ * and non-blockig mode.
+ *
+ * The provider is in charge of opening and closing file
+ * descriptors and connecting transport. After the connection has
+ * been established, all further IO operations will be delegated
+ * to the TLS service implementation which is provided by the application.
+ *
+ * The provider which is capable of using the service interface v1 must
+ * export the following function
+ *
+ * int wsrep_init_tls_service_v1(wsrep_tls_service_v1_t)
+ *
+ * which can be probed by the application.
+ *
+ * The application must initialize the service via above function
+ * before the provider is initialized.
+ */
+
+#ifndef WSREP_TLS_SERVICE_H
+#define WSREP_TLS_SERVICE_H
+
+#include <sys/types.h> /* posix size_t */
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif /* __cplusplus */
+
+/**
+ * TLS context type tag.
+ */
+typedef struct wsrep_tls_context_st wsrep_tls_context_t;
+
+/**
+ * TLS stream structure.
+ */
+typedef struct wsrep_tls_stream_st
+{
+    /**
+     * File descriptor corresponding to the stream. The provider is
+     * responsible in opening and closing the socket.
+     */
+    int fd;
+    /**
+     * Pointer to TLS context. For initializing valid TLS context see
+     * wsrep_tls_context_create_t below.
+     */
+    wsrep_tls_context_t* context;
+    /**
+     * Opaque pointer reserved for application use.
+     */
+    void* opaque;
```

Comment:
I take it, `wsrep_tls_context_t` is an opaque data type that is defined by the application. Why do we need one more `opaque` pointer?

Reply:
Context is global, `opaque` is per stream.

Reply:
I suppose we can have the context per stream and inside that context application can make pointers to global contexts if it needs to. From the rest of the API it does not look like provider needs that global context for anything as all operations are done on `wsrep_tls_stream_t` objects.